### PR TITLE
Solution: JavaScript III

### PR DESCRIPTION
**Pull request by nickcannariato**
## Git Set up

* [x] Fork the project into your GitHub user account
* [x] Clone the forked project into a directory on your machine
* [x] Create a pull request before you start working on the project requirements.  You will continuously push your updates throughout the project.
* [x] You are now ready to build this project with your preferred IDE

## Assignment Set Up

* Complete all the exercises as described inside each assignment file.
* To test your `console.log()` statements, open up the index.html file found in the assignments folder and use the developer tools to view the console.  

## The `this` keyword

Having a solid understanding of how `this` works will give you a huge advantage when you start building with more advanced frameworks. Use the [this.js](assignments/this.js) file to traverse through a few `this` problems.

## Prototype

The prototype challenge will focus on building prototypes for a fantasy themed game that includes mages, swordsmen, and archers.  Follow the [prototypes.js](assignments/this.js) instructions closely to create the beginnings of what could be an awesome JavaScript game.

* Read the instructions found within the file carefully to finish the challenges. 
* Remember to un-comment the objects and console logs to test your work at the bottom of the page.

### CODE REVIEW

File: assignments/prototypes.js [hunk 1]
```diff
@@ -14,6 +14,14 @@
...
   * destroy() // prototype method -> returns the string: 'Object was removed from the game.'
 */
+function GameObject(attrs) {
+  this.createdAt = attrs.createdAt
+  this.dimensions = attrs.dimensions
+};
+
+GameObject.prototype.destroy = function () {
+  return `Object was removed from the game`;
+}
```
Comment (sabrinagear) [inline assignments/prototypes.js]: Can you refactor this to make that last console.log() work?

File: assignments/prototypes.js [hunk 2]
```diff
@@ -22,6 +30,17 @@
...
+function CharacterStats(attrs) {
+  GameObject.call(this, attrs);
+  this.healthPoints = attrs.healthPoints;
+  this.name = attrs.name;
+}
+
+CharacterStats.prototype = Object.create(GameObject.prototype)
+
+CharacterStats.prototype.takeDamage = function () {
+  return `${this.name} took damage`;
```
Comment (sabrinagear) [inline assignments/prototypes.js]: Good job!

File: assignments/this.js [hunk 3]
```diff
@@ -1,26 +1,51 @@
...
 * in your own words. explain the four principle for the "this" keyword below.
 *
-* 1. 
-* 2. 
-* 3. 
-* 4. 
+* 1. Global -- When used in the global scope, `this` is bound to the Window or Console object, depending on where the code is being run
+* 2. Implicit -- When used in an object method, `this` is implicitly referring to the method's parent object. 
+* 3. New -- When used in the context of a constructor, `this` is referring to the specific instance that was created by the constructor. 
+* 4. Explicit--Whenever the `.call`, `.apply`, or `.bind` method is used, `this` will be explicitly defined.
```
Comment (sabrinagear) [inline assignments/this.js]: Good.

File: assignments/this.js [hunk 4]
```diff
@@ -1,26 +1,51 @@
...
 
 // Principle 4
 
-// code example for Explicit Binding
\ No newline at end of file
+// code example for Explicit Binding
+const michelle = new Person('Michelle', 29, 'She/Her'); // create a new person using the constructor function above
+
+console.log(jay.intro.call(michelle)); // Call jay object and with the michelle object's data.
+console.log(michelle.intro.apply(jay)); // Call michelle but explicitly apply jay object data to it.
```
Comment (sabrinagear) [inline assignments/this.js]: Good job!

File: assignments/prototypes.js (hunk 1 above)
Comment (nickcannariato) [reply assignments/prototypes.js]: @sabrinagear Yep! Committed to master since this was merged. 🙇

**Pull request merged**